### PR TITLE
[BEAM-14378] [CdapIO] SparkReceiverIO Read via SDF

### LINKED ISSUE #21680: [CdapIO] Design and implement SparkReceiverViaSDF component
### Context
SparkReceiverViaSDF component is supposed to be a part of the SparkReceiverIO connector and provide the functionality to process pipelines using a splittable DoFn approach
### Task Description
Required to design SparkReceiverViaSDF component and implement it. 
### Acceptance criteria
* Class diagram
* Sequence diagram 
* Source code

Imported from Jira [BEAM-14378](https://issues.apache.org/jira/browse/BEAM-14378). Original Jira may contain additional context.
Reported by: elizaveta.lomteva.

### PR DESCRIPTION
**Pull request by Amar3tto**
sdf-with-offset

Resolves #21680
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Add a link to the appropriate issue in your description, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run SQL_Java17 PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Lizzfox** [comment]
@aromanenko-dev 
Alexey, we've prepared the SparkReceiverIO connector Read interface. This PR is ready to be reviewed.
Thanks for your attention!

**Comment by aromanenko-dev** [comment]
Thanks!
I'm sorry for a delay with review, I'll try to take a look as soon as I can.

**Comment by Lizzfox** [comment]
Run Java CdapIO Performance Test

**Comment by Amar3tto** [comment]
@aromanenko-dev Kind ping on this. 
Did you have time to review this PR?

**Comment by aromanenko-dev** [comment]
Please, avoid `merge` commits and use `git rebase` instead

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java:37]: Please, add more details about this connector and examples of how to use it.

File: sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java:104]: No need to specify a runner - unit tests are running by default on DirectRunner

File: sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java:105]: Use `TestPipeline` for tests (see examples for other IO connectors).

File: sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java:123]: Use 
```
PCollection<String> output = p.apply(...)...

PAssert.that(output).containsInAnyOrder(expectedRecords);
```

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:123]: Better to avoid `@SuppressWarnings` if possible

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:118]: No need to specify the empty methods.

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:164]: Please, add more detailed error message, like `SparkReceiver was interrupted because of <>`.

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:171]: nit: `SparkReceiver is stopped.`

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:53]: Do we need this  class and its methods be `public`?

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:129]: Please, restrict methods visibility where it's possible.

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java:61]: Does this connector support portability?

**Review by aromanenko-dev — CHANGES_REQUESTED** [review]
Thanks for contribution!

I did a first round of code review (I didn't touch SDF part), ptal.

Also, please rebase your feature branch against `master` and squash your commits.

File: sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java]: Removed

File: sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java]: Done

File: sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java]: Done

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java]: Removed

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java]: Done

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java]: Done

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkConsumer.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkConsumer.java]: Changed to package-private

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java]: Changed to package-private

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java]: Done

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java:37]: We expect to add more detailed documentation and README in a separate PR.

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:118]: We are dealing with `org.apache.spark.streaming.receiver.ReceiverSupervisor` class which has `pushSingle(Object o)` method. So we need to do an unchecked cast from Object.
Should we add an explanation as a comment here?

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
Comment (aromanenko-dev) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java:37]: Is there any reason why not to add it along with this PR? Seems logical for me...

File: sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java
Comment (aromanenko-dev) [inline sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java:41]: Do you really need `PipelineOptions`? If not then it would be better create a pipeline in this way:
```
@Rule public TestPipeline readPipeline = TestPipeline.create();
```

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (aromanenko-dev) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:118]: Yes, please.

File: sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java:41]: Yes, we need to set `blockOnRun(false)`.

Comment (aromanenko-dev) [reply sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java:41]: What is a reason for that?

Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java:41]: We use `PipelineResult.waitUntilFinish()` following the example of others IO tests, and setBlockOnRun(false) is required in this case.

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java]: Added javadoc

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:120]: Done

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java]: No, removed annotation

**Comment by aromanenko-dev** [comment]
Thanks for addressing review comments. It looks ok in general for me. 

I'll be on PTO for the next 2 weeks, so I'd kindly ask @chamikaramj or @lukecwik to take a look on SDF part.

**Comment by chamikaramj** [comment]
Ack. Will take a look. Thanks @aromanenko-dev

**Comment by Amar3tto** [comment]
Run Java PreCommit

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:170]: At this moment all records for the current restriction should be already read from the queue.
Here we are stopping the Receiver and clearing the queue because it is not needed anymore.

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:151]: Yes it was intended, because it is `SerializableFunction<?, Void>` and it should return something. The common way is to return null here.

File: sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/CustomReceiverWithOffset.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/test/java/org/apache/beam/sdk/io/sparkreceiver/CustomReceiverWithOffset.java:67]: Fixed

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java]: Renamed

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java]: Renamed

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:154]: Done

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:162]: Done

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
Comment (Lizzfox) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java:158]: Could you please merge this implementation to start with, to unblock the work on PRs with Spark Recevier integration tests and integration with Cdap IO?
While we haven't found a way to split the Spark receivers, we continue to work in this direction

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (chamikaramj) [inline sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:198]: When the runner resume an element from a checkpoint, how will the source start reading from the correct position ?

I think there should be logic here to seek to the correct position when "currentRestriction().getFrom()" is not zero. 

Note that this has to be a seek (not read all elements from the beginning) for reading to be efficient.

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (chamikaramj) [inline sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:200]: I don't see any sort of acking here for messages that are already read. How does the SparkConsumer finalize messages so that it does not have to hold off to data forever ? Usually this is implemented via a bundle finalizer: https://beam.apache.org/documentation/programming-guide/#bundle-finalization

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
Comment (chamikaramj) [inline sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java:69]: Please add an experimental tag (see [1] for an example) since this is a new connector and the API can change.

[1] https://github.com/apache/beam/blob/f5435c0575870062f39575271c0f483117908403/sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java#L322

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
Comment (chamikaramj) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java:158]: Yeah, this can be pushed to a future PR. Let's add a TODO though.

**Review by chamikaramj — COMMENTED** [review]
Thanks. Added a couple more comments which I think are critical issues. Splitting can be pushed to a future change.

**Comment by chamikaramj** [comment]
Any updates ?

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java:72]: Done

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:198]: Start reading from the correct position ("seek") implemented here:
`sparkConsumer = new SparkConsumerWithOffset<>(tracker.currentRestriction().getFrom());`

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
Comment (Amar3tto) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java:200]: Discussed at the meeting  that an additional ack is not needed, since Spark Consumer is created inside DoFn. As soon as element is processed it’s deleted from queue (not from the source).

**Comment by Lizzfox** [comment]
@chamikaramj 
All your comments have been answered. The PR is ready to be merged.
Thank you.

**Review by chamikaramj — COMMENTED** [review]
Thanks. LGTM.

**Comment by Abacn** [comment]
New test flake seen on SparkReceiverIO test: #23449

**Pull request merged**
